### DISCUSSION TITLE
Running into issues outputting audio to Adafruit feather rp2040

### DISCUSSION THREAD

**Discussion by Dhoward094** [earlephilhower/arduino-pico#2587] [Q&A] [answered]
I am running into issues getting an audio output to a speaker connected to an Adafruit prop-maker feather rp2040. I have my code below that I flashed to the microcontroller, I get confirmation that the audio data is being received but no audio is played and I am unsure if I am outputting correctly.

```
#include <Arduino.h>
#include <I2S.h>

#define I2S_DATA 16 
#define I2S_BIT_CLOCK 17
#define I2S_WORD_SELECT 18

#define LED_PIN 13

#define SERIAL_BAUD_RATE 115200
#define AUDIO_BUFFER_SIZE 128

int16_t audioBuffer[AUDIO_BUFFER_SIZE];

I2S i2s(OUTPUT);

void setup() {
    pinMode(LED_PIN, OUTPUT);

    Serial.begin(SERIAL_BAUD_RATE);
    while (!Serial) {
        ;
    }

    i2s.setDATA(I2S_DATA);
    i2s.setBCLK(I2S_BIT_CLOCK);
    i2s.setBitsPerSample(16);
    i2s.setFrequency(44100);

    if (!i2s.begin()) {
        Serial.println("Failed to initialize I2S!");
        while (1);
    }
}

void loop() {
    if (Serial.available() >= AUDIO_BUFFER_SIZE * sizeof(int16_t)) {
        // Blink LED rapidly while receiving data
        digitalWrite(LED_PIN, HIGH);
        delay(167); // 3 times per second (1000 ms / 3 = ~333 ms cycle, 167 ms on/off)

        // Read the audio data into the buffer
        Serial.readBytes(reinterpret_cast<char*>(audioBuffer), AUDIO_BUFFER_SIZE * sizeof(int16_t));

        // Write the audio data to the I2S interface
        for (int i = 0; i < AUDIO_BUFFER_SIZE; i++) {
            i2s.write(audioBuffer[i]);
        }

        digitalWrite(LED_PIN, LOW);
        delay(167);
    } else {
        // Blink LED slowly while waiting for data
        digitalWrite(LED_PIN, HIGH);
        delay(500); // 1 time per second (500 ms on/off)

        digitalWrite(LED_PIN, LOW);
        delay(500);
    }
}
```

**Comment by earlephilhower** [comment, marked as answer]
You should try one of the I2S output examples to verify the pins.  If the examples don't work then the pins you're using may not be right, or maybe the I2S DAC needs some kind of control to actually turn on (i.e. does it need an I2C write or GPIO pin set high?).

I'm not sure the USB CDC buffer can actually grow to 128 bytes  so your loop may never actually read anything.  We don't use a separate FIFO since it would just be wasted memory copies.  So, try 16 or 32 bytes instead of 128 in your buffer read/write size.

Also, while 44.1Khz, 16b, stereo only requires ~1.4Mbit and USB 1.1 theoretically gives 12Mbit, I think you're not actually going to be able to reach that rate using a normal CDC(Serial) interface.  To get high rates I believe isochronous channels need to be set up (like a USB audio card) to guarantee bandwidth and latency.  That's beyond the scope of this core.  You may want to consider only doing mono (and duplicating samples in the code), or 16Khz (and changing the I2s frequency), or even 8-bit writes (rehydrate to 16b in code).

No matter what you do, you'll probably want to increase the i2s buffer size to help avoid underflow once things get started.  This is in the examples as well.

**Reply by Dhoward094** [reply]
Okay, this sets me in the right direction. Thank you for your help!
